File: wallet/src/de/schildbach/wallet/AppDatabase.java
package de.schildbach.wallet;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;

import org.dash.wallet.common.data.AddressMetadata;
import org.dash.wallet.common.data.BlockchainState;
import org.dash.wallet.common.data.ExchangeRate;
import org.dash.wallet.common.data.RoomConverters;
import org.dash.wallet.common.data.TransactionMetadata;

import de.schildbach.wallet.data.AddressMetadataDao;
import de.schildbach.wallet.data.AppDatabaseMigrations;
import de.schildbach.wallet.data.BlockchainIdentityData;
import de.schildbach.wallet.data.BlockchainIdentityDataDao;
import de.schildbach.wallet.data.BlockchainStateDao;
import de.schildbach.wallet.data.BlockchainStateRoomConverters;
import de.schildbach.wallet.data.DashPayContactRequest;
import de.schildbach.wallet.data.DashPayContactRequestDao;
import de.schildbach.wallet.data.DashPayProfile;
import de.schildbach.wallet.data.DashPayProfileDao;
import de.schildbach.wallet.data.Invitation;
import de.schildbach.wallet.data.InvitationsDao;
import de.schildbach.wallet.data.TransactionMetadataChangeCacheDao;
import de.schildbach.wallet.data.TransactionMetadataCacheItem;
import de.schildbach.wallet.data.TransactionMetadataDocument;
import de.schildbach.wallet.data.TransactionMetadataDao;
import de.schildbach.wallet.data.TransactionMetadataDocumentDao;
import de.schildbach.wallet.data.UserAlertDao;
import de.schildbach.wallet.rates.ExchangeRatesDao;
import de.schildbach.wallet.ui.dashpay.UserAlert;
import kotlin.Deprecated;

/**
 * @author Samuel Barbosa
 */
@Database(entities = {
        ExchangeRate.class,
        BlockchainState.class,
        TransactionMetadata.class,
        AddressMetadata.class,
        BlockchainIdentityData.class,
        DashPayProfile.class,
        DashPayContactRequest.class,
        UserAlert.class,
        Invitation.class,
        TransactionMetadataCacheItem.class,
        TransactionMetadataDocument.class
}, version = 18)
@TypeConverters({RoomConverters.class, BlockchainStateRoomConverters.class})
public abstract class AppDatabase extends RoomDatabase {

    private static AppDatabase instance;

    public abstract ExchangeRatesDao exchangeRatesDao();
    public abstract BlockchainStateDao blockchainStateDao();
    public abstract TransactionMetadataDao transactionMetadataDao();
    public abstract AddressMetadataDao addressMetadataDao();

    public abstract BlockchainIdentityDataDao blockchainIdentityDataDao();

    public abstract DashPayProfileDao dashPayProfileDao();

    public abstract DashPayContactRequestDao dashPayContactRequestDao();

    public abstract InvitationsDao invitationsDao();

    public abstract TransactionMetadataChangeCacheDao transactionMetadataCacheDao();

    public abstract TransactionMetadataDocumentDao transactionMetadataDocumentDao();

    public abstract UserAlertDao userAlertDao();

    @Deprecated(message = "Inject instead")
    public static AppDatabase getAppDatabase() {
        if (instance == null) {
            // destructive migrations are used from versions 1 to 10
            instance = Room.databaseBuilder(WalletApplication.getInstance(),
                    AppDatabase.class, "dash-wallet-database")
                    .addMigrations(
                            AppDatabaseMigrations.getMigration11To17(),
                            AppDatabaseMigrations.getMigration16To17(),
                            AppDatabaseMigrations.getMigration17To18()
                    )
                    .fallbackToDestructiveMigrationFrom(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).build();
        }
        return instance;
    }

}

File: wallet/src/de/schildbach/wallet/database/AppDatabase.kt
/*
 * Copyright 2023 Dash Core Group.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.schildbach.wallet.database

import androidx.room.Database
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import de.schildbach.wallet.database.dao.*
import de.schildbach.wallet.database.entity.DashPayContactRequest
import de.schildbach.wallet.database.entity.DashPayProfile
import de.schildbach.wallet.database.entity.ImportedMasternodeKey
import de.schildbach.wallet.database.entity.Invitation
import de.schildbach.wallet.database.entity.TopUp
import de.schildbach.wallet.database.entity.TransactionMetadataCacheItem
import de.schildbach.wallet.database.entity.TransactionMetadataDocument
import de.schildbach.wallet.database.entity.UsernameRequest
import de.schildbach.wallet.database.entity.UsernameVote
import de.schildbach.wallet.ui.dashpay.UserAlert
import org.dash.wallet.common.data.entity.AddressMetadata
import org.dash.wallet.common.data.entity.BlockchainState
import org.dash.wallet.common.data.entity.ExchangeRate
import org.dash.wallet.common.data.entity.IconBitmap
import org.dash.wallet.common.data.entity.TransactionMetadata
import org.dash.wallet.features.exploredash.data.dashdirect.GiftCardDao
import org.dash.wallet.common.data.entity.GiftCard
import org.dash.wallet.features.exploredash.utils.RoomConverters

@Database(
    entities =
    [
        ExchangeRate::class,
        BlockchainState::class,
        TransactionMetadata::class,
        AddressMetadata::class,
        IconBitmap::class,
        GiftCard::class,
        DashPayProfile::class,
        DashPayContactRequest::class,
        UserAlert::class,
        Invitation::class,
        TransactionMetadataCacheItem::class,
        TransactionMetadataDocument::class,
        UsernameRequest::class,
        UsernameVote::class,
        ImportedMasternodeKey::class,
        TopUp::class
    ],
    version = 15, // if increasing version, we need migrations to preserve tx/addr metadata,
    exportSchema = true
)
@TypeConverters(RoomConverters::class, BlockchainStateRoomConverters::class)
abstract class AppDatabase : RoomDatabase() {
    abstract fun exchangeRatesDao(): ExchangeRatesDao
    abstract fun blockchainStateDao(): BlockchainStateDao
    abstract fun transactionMetadataDao(): TransactionMetadataDao
    abstract fun addressMetadataDao(): AddressMetadataDao
    abstract fun iconBitmapDao(): IconBitmapDao
    abstract fun giftCardDao(): GiftCardDao
    abstract fun dashPayProfileDao(): DashPayProfileDao
    abstract fun dashPayContactRequestDao(): DashPayContactRequestDao
    abstract fun invitationsDao(): InvitationsDao
    abstract fun transactionMetadataCacheDao(): TransactionMetadataChangeCacheDao
    abstract fun transactionMetadataDocumentDao(): TransactionMetadataDocumentDao
    abstract fun userAlertDao(): UserAlertDao
    abstract fun usernameRequestDao(): UsernameRequestDao
    abstract fun usernameVoteDao(): UsernameVoteDao
    abstract fun importedMasternodeKeyDao(): ImportedMasternodeKeyDao
    abstract fun topUpsDao(): TopUpsDao
}
